### PR TITLE
Improve project management

### PR DESCRIPTION
**Pull request by JLDLaughlin**
As the EPD organization at Materialize has grown, each team has started to develop their own project management processes. We've gotten to the size where this divergence limits cross-team and cross-company understanding about what we're working on, why we're working on it, and when folks can expect any piece of work to be available for customer use. Additionally, our informal processes haven't frequently included an important step: publicizing the things that we're working on.

This PR attempts to outline a new joint Engineering and Product project management process. This process will replace any existing Engineering and Product processes, and will dictate how we prioritize and deliver work moving forward.

I'd recommend reviewing this PR in the following order:

- `project-management.md`: this is the highest-level document, and provides context for the other changes.
- `design/README.md`: this is fairly similar to the existing Engineering design process, but with a bit of a Product twist.
- `design/00000000_template.md`: this is an update to the existing design document template, and also replaces the Compute-specific one!
- `feature-lifecycle.md`: once a project is out of the Design phase, this document explains how we'll be managing the development and rollout of projects.

I don't think we'll get to complete consensus here, but I'm looking to sand down any rough edges! Let me know what looks good, what doesn't, and what I may have missed.

### CODE REVIEW

File: doc/developer/design/README.md
Comment (aalexandrov) [inline doc/developer/design/README.md:136]: It's hard for me to reconcile

> If you did not hear back from important stakeholders, do not take this as approval. Make sure the relevant people have in fact seen your design and had a chance to object or propose changes.

with 

> Design document meetings are not meant for open-ended brainstorming, nor are they meant to act as a review body. The intention of this process is not to design by committee.

At some point I might want to merge a design doc without unanimous consensus. Are there specific guidelines that I should adhere before doing so (for example, record conflicting opinions in the design doc for posterity).

File: doc/developer/feature-lifecycle.md
Comment (aalexandrov) [inline doc/developer/feature-lifecycle.md:5]: what is DRI?

File: doc/developer/feature-lifecycle.md
Comment (nrainer-materialize) [inline doc/developer/feature-lifecycle.md:106]: Should we discuss at this stage whether the feature flag shall stay or be removed?

File: doc/developer/design/00000000_template.md
Comment (rjobanp) [inline doc/developer/design/00000000_template.md:64]: is this always a prerequisite for creating a new design doc? If not, should it be clarified that this is an optional and/or secondary step after the initial proposal has some eyes on it?

File: doc/developer/design/00000000_template.md
Comment (rjobanp) [inline doc/developer/design/00000000_template.md:52]: since it looks like we've combined the `Overview` and `Detailed description` blocks into this `Solution Proposal` block, suggesting the author include a brief summary/high-level description at the start of the proposal could be helpful for readers to get a quick understanding before reading all the technical details

**Review by rjobanp — COMMENTED** [review]
nice work!

File: doc/developer/design/00000000_template.md
Comment (moulimukherjee) [inline doc/developer/design/00000000_template.md:34]: I personally like the idea of an Exit Criteria rather than a Success Criteria which is a subset. Once the implementation is underway, we might learn more which might change or limit what we hoped to achieve. Being cognizant about the risks and the potential end results is helpful I think.

**Comment by parkerhendo** [comment]
@JLDLaughlin relaying something that came up in the Console team meeting today

> At what point in the feature lifecycle does a feature's implementation get "locked in" to ensure there's no breaking changes to it going forward? Secondly, at what point in the lifecycle do new "features"/tables move from `mz_internal` to `mz_catalog`?

cc: @chaas

File: doc/developer/design/README.md
Comment (aalexandrov) [reply doc/developer/design/README.md:76]: > Does it make sense to note down here that the design document should be written and completed before starting to work on the code (so that design doc changes resulting from a discussion don't require reworking the code)?

I wouldn't say "completed" but "reviewed" or "discussed" in that case.In my mind a design doc should not be considered "completed" until the code that implements it is written.

Asking to have a design doc up until the point where all unknowns are resolved before we can start with the implementation means failing to acknowledge that for almost all design docs we are going to encounter unknown unknowns when we start working on the coding part.

We can (and probably should) merge the doc after the initial round of reviews, but we should make sure that the content of the design doc is aligned with the code that realizes it.

File: doc/developer/project-management.md
Comment (teskje) [inline doc/developer/project-management.md:108]: Does that mean all projects that warrant a design doc are supposed to go through the feature lifecycle? ISTM that the feature lifecycle can only be sensibly applied to large user-facing features (considering that it includes requirements like external testing and polished documentation, a blog post, ...), but we will also want to write design docs for internal changes.

File: doc/developer/project-management.md
Comment (bkirwi) [inline doc/developer/project-management.md:61]: Is triage the process that assigns these priorities? (Does this imply that all issues are triaged within 24h in case one of them is P0?)

File: doc/developer/design/README.md
Comment (bkirwi) [reply doc/developer/design/README.md:76]: I agree with Alexander on this one. Speaking personally, the design docs I've worked on would be much lower quality if we hadn't done some implementation work first. (And given the process here, "fixing" a design doc is pretty expensive... it needs to go through the same set of reviews as the original doc. That means it's better to spend a little time to get it right up front!)

There's definitely _some_ line, though. Maybe it's fair to say that the design docs should land before the proposed feature becomes customer-visible? So prototyping on a branch or behind a flag would be fine.

**Comment by chaas** [comment]
> @JLDLaughlin relaying something that came up in the Console team meeting today
> 
> > At what point in the feature lifecycle does a database feature's implementation/design get "locked in" to ensure there's no breaking changes to it going forward? Secondly, at what point in the lifecycle do new "features"/tables move from `mz_internal` to `mz_catalog`?
> 
> cc: @chaas

I had another thought in this vein. What happens when a feature has an "extended stay" in private preview. Does it stay on the tracking board for months as we wait for feedback? If not, where/how do we broadcast the upcoming transition? 
A recent example is the subscribe output formats. We implemented it a few months ago and "closed the epic", but only recently got enough usage data to decide to move one of the formats (envelope upsert) to public preview. Where should we announce/track that?

File: doc/developer/project-management.md
Comment (antiguru) [inline doc/developer/project-management.md:106]: I'm not sure if this will be successful. For many changes, we need to work on the design in parallel to the implementation because there is no other way to figure out what question the design doc needs to answer. How does this fit within this proposal?

File: doc/developer/project-management.md
Comment (antiguru) [reply doc/developer/project-management.md:108]: I agree, we should clarify what changes should go through this process. Is it all, or just those tracked by EPD?

File: doc/developer/feature-lifecycle.md
Comment (joacoc) [inline doc/developer/feature-lifecycle.md:76]: Should projects like `terraform`, `dbt`, and `mz` include metrics and observability? I am inclined towards collecting metrics and enabling observation (e.g. Sentry), although there might be cases where it's not feasible or excessive.

File: doc/developer/design/00000000_template.md
Comment (uce) [reply doc/developer/design/00000000_template.md:52]: Added a note in 0d4e2f7488.

File: doc/developer/design/README.md
Comment (uce) [reply doc/developer/design/README.md:134]: I added a note in 5d62e93dde to record meeting outcomes.

File: doc/developer/design/README.md
Comment (uce) [reply doc/developer/design/README.md:138]: The important bit is that relevant stakeholders were involved without formally requiring their consensus.

This part above is relevant:

> Instead, the author is responsible for driving the process to a conclusion that is satisfactory to the relevant stakeholders. Not every design document will result in complete consensus.

Regarding guidlines: I added a note in 5d62e93dde to record outcomes of the meeting. That should make it clearer whether there was consensus on a design or not.

File: doc/developer/design/00000000_template.md
Comment (uce) [reply doc/developer/design/00000000_template.md:13]: I don't see the note at conflict with the first paragraph. The recommendation is to follow the suggested structure while highlighting that answering the respective questions is the important bit. I like that we're providing flexibility to deviate from the structure as long as the main questions are answered.

I'll resolve the comment, but feel free to reopen if you feel strongly about this. I think the main option we'd have is to remove the note.

File: doc/developer/design/00000000_template.md
Comment (uce) [reply doc/developer/design/00000000_template.md:34]: If the above works for you, I'd love to resolve this comment.

File: doc/developer/design/00000000_template.md
Comment (uce) [reply doc/developer/design/00000000_template.md:61]: The question of how much detail to provide for the solution proposal did pop up a few times in the past. If you don't mind, I'd err on the side of having too many hints in the doc rather than too few. Having an explicit prompt like this can make it more comfortable to ask for help or feedback.

File: doc/developer/design/00000000_template.md
Comment (moulimukherjee) [reply doc/developer/design/00000000_template.md:34]: Sounds good 👍

File: doc/developer/design/README.md
Comment (uce) [reply doc/developer/design/README.md]: Fixed in ee094d47ee.

File: doc/developer/design/README.md
Comment (uce) [reply doc/developer/design/README.md]: Added link to template in 0ec3ce9555. The link to design was removed, because the documentation here applies to other repos, too (e.g. in cloud we have doc/design whereas here we have doc/developer/design).

File: doc/developer/feature-lifecycle.md
Comment (uce) [reply doc/developer/feature-lifecycle.md]: Spelled out in 0339898be4.

File: doc/developer/project-management.md
Comment (uce) [reply doc/developer/project-management.md:39]: "Problems can identify potential improvements as well as net-new features."

The idea with calling it "problem" is that we want to focus on the problem we're solving rather than the solution.

File: doc/developer/project-management.md
Comment (uce) [reply doc/developer/project-management.md]: Mentioned 7a110f5b47.

File: doc/developer/project-management.md
Comment (uce) [reply doc/developer/project-management.md]: No, typically the issue filer assigns the `p0` or `release-blocker` label. The goal of the triage process is to ensure that every issue filed issue against our main repositories gets processed (in a timely manner).

Tried to clarify in 30970683ac.

File: doc/developer/project-management.md
Comment (uce) [reply doc/developer/project-management.md:111]: I think Moritz' framing as "tracked by EPD" makes sense. Having said that, that will still apply to a majority of projects in my expectation. We definitely have room in this framework for design documents for internal changes that are not tracked by EPD though.

File: doc/developer/project-management.md
Comment (uce) [reply doc/developer/project-management.md:109]: The discovery phase explicitly asks for prototypes. It's totally fine if they evolve together with the design. I think your work on unbilled replicas is a good example. The design and prototype evolved hand in hand and once the design is approved you will deliver based on the prototype – it's totally fine if the diff between the prototype and the final delivery is small.

File: doc/developer/project-management.md
Comment (uce) [reply doc/developer/project-management.md]: Yes, there is no special triage expectation for those labels (beyond what's written in the doc). We will update the label descriptions (e.g. they currently speak to addressing bugs for the next release cycle). If an individual team finds them valuable, I don't mind keeping them around.

File: doc/developer/project-management.md
Comment (uce) [reply doc/developer/project-management.md:72]: The paragraph is mainly saying that it's fine to short-circuit the process when the issue is straightforward. I think that's different from the problem (which I agree with) you are describing. I'll note this down for a follow-up, but wouldn't want to address as part of this PR.

I did a quick check and we have [300+ assigned issues](https://github.com/MaterializeInc/materialize/issues/assigned/*). A large fraction of those are lingering assignments (e.g. dating back to 2019) and a smaller fraction is meaningful (e.g. self-assigned to fix).

File: doc/developer/feature-lifecycle.md
Comment (uce) [reply doc/developer/feature-lifecycle.md:88]: We have a list of criteria below–including announcements and docs, but not limited to. The key point is that public preview is about ensuring stability (including performance). Another way to look at is as beta ~ public preview vs. stable ~ GA.

File: doc/developer/feature-lifecycle.md
Comment (uce) [reply doc/developer/feature-lifecycle.md:77]: I'll resolve this comment, but please raise this with @morsapaes if it is unclear. She can drive the resolution forward on this.

File: doc/developer/feature-lifecycle.md
Comment (uce) [reply doc/developer/feature-lifecycle.md:68]: That's the inherent risk associated with using private preview features. There is no magic bullet here and we'll likely have to treat it on a case-by-case basis until we have a few precedents for this (we don't have any yet).

File: doc/developer/feature-lifecycle.md
Comment (uce) [reply doc/developer/feature-lifecycle.md:37]: That's exactly what the document is saying. **Before** moving to private preview, there need to be stash migrations if needed. The document does not say that these things have to happen at the end of the internal development phase.

File: doc/developer/feature-lifecycle.md
Comment (uce) [reply doc/developer/feature-lifecycle.md:15]: I'll note down to rename #eng-announce to #epd-announce (or spin it up as a separate channel.

File: doc/developer/design/README.md
Comment (uce) [reply doc/developer/design/README.md]: Correct. The #eng-design-docs channel allows everyone to follow what design documents are in flight, but that's different from someone asking for additional input in #epd-announce. I added a note in 42c9035fd2 and removed the recommendation in 5b93ac4463.

File: doc/developer/design/00000000_template.md
Comment (uce) [reply doc/developer/design/00000000_template.md:67]: I wouldn't want to call it optional, because it is such a valuable step, but I added a note in 8f7dae5cab to leave some room for doing it at a later time (or skipping it) when there is a good reason for it. @JLDLaughlin wdyt?

File: doc/developer/design/README.md
Comment (uce) [reply doc/developer/design/README.md:54]: Added to the template in 8b8694b846.

File: doc/developer/design/README.md
Comment (uce) [reply doc/developer/design/README.md:78]: Agree! Added a note on prototypes/experimental work in 18e5a1148b.

File: doc/developer/feature-lifecycle.md
Comment (uce) [reply doc/developer/feature-lifecycle.md]: Added in c358e76750.

File: doc/developer/feature-lifecycle.md
Comment (uce) [reply doc/developer/feature-lifecycle.md:112]: Added in c358e76750.

File: doc/developer/feature-lifecycle.md
Comment (uce) [reply doc/developer/feature-lifecycle.md:14]: Added in c358e76750.

File: doc/developer/feature-lifecycle.md
Comment (uce) [reply doc/developer/feature-lifecycle.md]: I added more details in ba1a6e741f. This was a good opportunity to add another note to reach out to your manager if you're unsure. 🙈

File: doc/developer/feature-lifecycle.md
Comment (uce) [reply doc/developer/feature-lifecycle.md:30]: > Alternatively, should we work towards making it possible to feature-gate all features?

Yes, we should push to make it easier when we encounter places where it's hard (e.g.  we did this recently for Brennan's encoding change).

I'm not too worried about minor leakage as long as it doesn't affect the stability of customer environments and we don't paint ourselves into a corner. I think the listed exit criteria are a good reminder to have those discussions on a case-by-case basis.

File: doc/developer/feature-lifecycle.md
Comment (uce) [reply doc/developer/feature-lifecycle.md:71]: Agreed. I'm noting this down as a follow-up.

Comment (uce) [reply doc/developer/feature-lifecycle.md:71]: Let me actually remove the reference to the doc, because it's confusing.

Comment (uce) [reply doc/developer/feature-lifecycle.md]: Replaced with link to feature flagging docs in 5892759c63.

File: doc/developer/design/00000000_template.md
Comment (JLDLaughlin) [reply doc/developer/design/00000000_template.md:67]: I think we can leave room to skip it if there's a good reason for it. In most cases, though, I'll expect to see something here!

Comment (uce) [reply doc/developer/design/00000000_template.md:67]: Thanks, Jessica. Updated in 63159579ac.

File: doc/developer/feature-lifecycle.md
Comment (uce) [reply doc/developer/feature-lifecycle.md]: I renamed feature lifecycle to project lifecycle in 2e1a5a3d0e. I think that's consistent with the overall "project management" framing and has all the right connotations.

To answer @rjobanp's question: yes, by default the process should apply to all processes.

File: doc/developer/project-management.md
Comment (uce) [reply doc/developer/project-management.md]: EPD management is responsible and can delegate to team members.

**Comment by uce** [comment]
Thanks everyone who reviewed the changes. The feedback was helpful in sanding down rough edges. As Jessica noted it's unlikely that we'll get to complete consensus on this. If a comment wasn't addressed to your satisfaction, please feel free to reach out to me or Jessica.

Regarding @chaas's question about how long we keep GitHub epics open: the idea is that  epics act as trackers that stay open until the project reaches GA. In the subscribe output format example, the respective epic would be still open in private preview and we'd set its status accordingly and keep it on the EPD backlog.

I'd like to merge the changes here after approval from @JLDLaughlin.

I have noted down the following follow-ups:
- [ ] Update issue templates, including for epics
- [ ] Change prefix of Slack channels from `eng-` to `epd-`
- [ ] Update documentation for feature flag lifecyle
- [ ] Update documentation for triage (what does assignment mean, etc.)
- [ ] Update description of P1-P3 labels
- [ ] Add project management docs to other repos (linking here)

**Review by benesch — APPROVED** [review]

**Pull request merged**
